src/search/command_line.cc

```diff
@@ -184,6 +184,118 @@ shared_ptr<SearchAlgorithm> parse_cmd_line(
+        suggestions.push_back("--if-non-unit-cost");
+        suggestions.push_back("--always");
+        // remove suggestions not starting with current_word
+    }
+
+    if (!current_word.empty()) {
+        // Suggest only words that match with current_word
+        suggestions.erase(
+            remove_if(suggestions.begin(), suggestions.end(),
+                      [&](const string &value) {
+                          return !value.starts_with(current_word);
+                      }), suggestions.end());
+    }
+    return suggestions;
+}
+
+void handle_tab_completion(int argc, const char **argv) {
+    if (argc < 2 || static_cast<string>(argv[1]) != "--bash-complete") {
+        return;
+    }
+    if (argc < 5) {
+        input_error(
+            "The option --bash-complete is only meant to be called "
+            "internally to generate suggestions for tab completion.\n"
+            "Usage:\n    ./downward --bash-complete "
+            "$COMP_POINT \"$COMP_LINE\" $COMP_CWORD ${COMP_WORDS[@]}\n"
+            "where the environment variables have their usual meaning for bash completion:\n"
+            "$COMP_POINT is the position of the cursor in the command line.\n"
+            "$COMP_LINE is the current command line.\n"
+            "$COMP_CWORD is an index into ${COMP_WORDS} of the word under the cursor.\n"
+            "$COMP_WORDS is the current command line split into words.\n"
+            );
+    }
+    int cursor_pos = parse_int_arg("COMP_POINT", static_cast<string>(argv[2]));
+    const string command_line = static_cast<string>(argv[3]);
+    int cursor_word_index = parse_int_arg("COMP_CWORD", static_cast<string>(argv[4]));
+    vector<string> words;
+    words.reserve(argc - 4);
+    for (int i = 5; i < argc; ++i) {
+        words.emplace_back(argv[i]);
+    }
+    // Sentinal for cases where the cursor is after the last word.
+    words.push_back("");
+
+    vector<string> parsed_args;
+    string current_word;
+    int pos = 0;
+    int end = static_cast<int>(command_line.size());
+    int num_words = static_cast<int>(words.size());
+    for (int i = 0; i < num_words; ++i) {
+        current_word = words[i];
+        int word_len = static_cast<int>(current_word.size());
+        if (command_line.substr(pos, word_len) != current_word) {
+            input_error("Expected '" + current_word + "' in $COMP_LINE at position "
+                        + to_string(pos));
+        }
+        if (i == cursor_word_index) {
+            if (cursor_pos < 0 || cursor_pos > word_len) {
+                input_error("Inconsistent information in COMP_LINE and COMP_WORDS");
+            }
+            break;
+        }
+
+        // Skip word
+        pos += word_len;
+        cursor_pos -= word_len;
+        parsed_args.push_back(current_word);
+
+        // Skip whitespace between words.
+        while (pos < end && isspace(command_line[pos])) {
+            ++pos;
+            --cursor_pos;
+        }
+    }
```

Comment:
I don't fully understand the code, but is it OK that the test of the consistent cursor_pos only happens when we break out of the loop, and there is no test if the loop completes without breaking out?

Reply:
With the new code, the test is now always performed.